src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                            string localFunctionName =_semanticFacts.GenerateNameForExpression(
+                                _semanticModel, 
+                                source, 
+                                capitalize: false, 
+                                _cancellationToken);
+                            localFunctionName = GetFreeSymbolName(localFunctionName, source.GetLocation().SourceSpan.Start);
+                            var localFunctionDeclaration = SyntaxFactory.LocalFunctionStatement(
+                                modifiers: default,
+                                returnType: typeSyntax.WithAdditionalAnnotations(Simplifier.SpecialTypeAnnotation),
+                                identifier: SyntaxFactory.Identifier(localFunctionName),
+                                typeParameterList: null,
+                                parameterList: SyntaxFactory.ParameterList(),
+                                constraintClauses: default,
+                                body: SyntaxFactory.Block(
+                                    SyntaxFactory.Token(
+                                        SyntaxFactory.TriviaList(),
+                                        SyntaxKind.OpenBraceToken,
+                                        SyntaxFactory.TriviaList(SyntaxFactory.EndOfLine(Environment.NewLine))),
+                                    SyntaxFactory.SingletonList(convertedFromQueryExpression),
+                                    SyntaxFactory.Token(SyntaxKind.CloseBraceToken)),
+                                expressionBody: null);
+
+                            var localFunctionInvocation = SyntaxFactory.InvocationExpression(SyntaxFactory.IdentifierName(localFunctionName)).WithAdditionalAnnotations(Simplifier.Annotation);
+                            SyntaxNode newParentExpressionStatement = parentStatement.ReplaceNode(source, localFunctionInvocation.WithAdditionalAnnotations(Simplifier.Annotation));
+                            documentUpdateInfo = new DocumentUpdateInfo(parentStatement, new[] { localFunctionDeclaration, newParentExpressionStatement });
+                            return true;
+                        }
+                    }
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool TryGetTypeSyntax(ExpressionSyntax expression, out TypeSyntax typeSyntax)
+            {
+                var typeSymbol = _semanticModel.GetTypeInfo(expression, _cancellationToken).Type;
+                if (typeSymbol.TypeKind == TypeKind.Error ||
+                    typeSymbol.ContainsAnonymousType())
+                {
+                    typeSyntax = default;
+                    return false;
+                }
+
+                typeSyntax = typeSymbol.GenerateTypeSyntax();
+                return true;
+            }
+
+            private string GetFreeSymbolName(string prefix, int position)
```

Comment:
>GetFreeSymbolName [](start = 27, length = 17)

Is this function going to avoid conflict with names used in nested scopes?

Reply:
Replaced it with _semanticFacts.GenerateUniqueName(_semanticModel, _source, containerOpt: null, baseName: prefix, _cancellationToken);
Added tests

---
In reply to: [176947151](https://github.com/dotnet/roslyn/pull/25362#discussion_r176947151) [](ancestors = 176947151)